server/iframe.go

```diff
@@ -72,16 +76,49 @@ func (a *API) authenticate(w http.ResponseWriter, r *http.Request) {
 	var logger logrus.FieldLogger
 	logger = logrus.StandardLogger()
 
+	redirectPath := "/"
+
+	// Check if we have a subEntityID coming from the Microsoft Teams SDK to redirect the user to the correct URL.
+	// We use this from the Team's notifications to redirect the user to what triggered the notification, in this case,
+	// a post.
+	subEntityID := r.URL.Query().Get("sub_entity_id")
+	if subEntityID != "" {
+		if strings.HasPrefix(subEntityID, "post_") {
+			postID := strings.TrimPrefix(subEntityID, "post_")
+			post, err := a.p.API.GetPost(postID)
+			if err != nil {
+				logger.WithError(err).Error("Failed to get post to generate redirect path from subEntityId")
+			}
+
+			channel, appErr := a.p.API.GetChannel(post.ChannelId)
+			if appErr != nil {
+				logger.WithError(appErr).Error("Failed to get channel to generate redirect path from subEntityId")
+			}
+
+			team, appErr := a.p.API.GetTeam(channel.TeamId)
+			if appErr != nil {
+				logger.WithError(appErr).Error("Failed to get team to generate redirect path from subEntityId")
+			}
+
+			redirectPath = fmt.Sprintf("/%s/pl/%s", team.Name, postID)
+		}
+	}
+
 	// If the user is already logged in, redirect to the home page.
 	// TODO: Refactor the user properties setup to a function and call it from here if the user is already logged in
 	// just in case the user logs in from a tabApp in a browser.
 	if r.Header.Get("Mattermost-User-ID") != "" {
 		logger = logger.WithField("user_id", r.Header.Get("Mattermost-User-ID"))
 		logger.Info("Skipping authentication, user already logged in")
-		http.Redirect(w, r, "/", http.StatusSeeOther)
+		http.Redirect(w, r, redirectPath, http.StatusSeeOther)
 		return
 	}
 
+	appID := r.URL.Query().Get("app_id")
+	if appID == "" {
```

Comment:
I think this can be closer to where it is used.  That will help remove any confusion about why this error is logged but no action taken.